core/src/main/java/org/apache/iceberg/TableMetadataParser.java

```diff
@@ -288,6 +292,9 @@ static TableMetadata fromJson(FileIO io, InputFile file, JsonNode node) {
           schema, TableMetadata.INITIAL_SPEC_ID, node.get(PARTITION_SPEC)));
     }
 
+    int lastAssignedFieldId = Optional.ofNullable(JsonUtil.getIntOrNull(LAST_ASSIGNED_FIELD_ID, node))
+        .orElseGet(() -> specs.stream().mapToInt(PartitionSpec::lastAssignedFieldId).max().orElse(999));
```

Comment:
Do we want it to be a required field in v2 and fail validation if it's missing?

Reply:
It makes sense as the V2 is not released yet. I will update it accordingly.